kloud: use metadata for deploying klient

rjeczalik opened the pull request:
This PR wires kloud/metadata with stack/provider, allowing kloud to deploy a klient service without a .deb builder dependency. Since the metadata provisioning method is generic, it allows kloud to build stacks on images other than Ubuntu 14.04.

The following distributions were tested:

- RedHat 7.3
- Suse 12 SP2
- Ubuntu 16.04
- Ubuntu 14.04
- Amazon Linux

![distros](http://g.recordit.co/EUQ8t7CVaX.gif)

A big todo here is to create here terraform tests, which could be integrated with CI to test such stacks. This is an ongoing effort to introduce integration testing for kloud, but outside the scope of this PR. I'm planning to fix it after finalising kd.io - then use kd for the testing.

Blockers:

- ~https://github.com/koding/service/issues/6~

rjeczalik committed:
kloud/metadata: make provision.sh a static asset

Instead of templating the provision.sh script, pass
the default configuration to it via terraform's
variable block.

The variables can be overwritten via environmental
variables.

rjeczalik committed:
kloud/metadata: add New for creating cloud-init from user-data

rjeczalik committed:
stack/provider: use metadata for provisioning klient

rjeczalik committed:
kloud/metadata: add inline doc

rjeczalik commented:
Also it will be possible to use custom cloud-init script for the user_data field, so e.g. instead of injecting ssh keys bash-fu, one can just create the following stack template:

```yaml
provider:
  aws:
    access_key: '${var.aws_access_key}'
    secret_key: '${var.aws_secret_key}'
resource:
  aws_instance:
    aws-instance:
      instance_type: 't2.nano'
      ami: ''
      tags:
        Name: '${var.koding_user_username}-${var.koding_group_slug}'
      user_data: |
        #cloud-config
        ssh_authorized_keys:
          - ssh-rsa AAAAB3NzaC1yc2EA... koding-350298856
```

/cc @gokmen

rjeczalik committed:
klient: ensure CurrentUser is read after it's written by flag.Parse

rjeczalik committed:
klient: bump go version for builds

rjeczalik committed:
client: remove comments about Ubuntu 14.04 restrictions

rjeczalik committed:
kloud/metadata: replace wget with curl

rjeczalik committed:
kloud/metadata: ensure /var/log/klient.log is owned by user

rjeczalik committed:
klient/terminal: rework screen command exec

rjeczalik committed:
klient/terminal: fix user / env handling

rjeczalik committed:
kloud/metadata: ensure proper perms for screen dir

rjeczalik committed:
kloud: use native screen when available

Embedded screen seems not to work on Red Hat.

rjeczalik committed:
stack/provider: simplify metadata handling

rjeczalik committed:
kloud/metadata: ensure cloud-init-output.log has proper perms

rjeczalik committed:
kloud/metadata: fix provision script termination

Not all cloud-inits implementations support final_message directive,
yield end of provisioning phase within the provision.sh script instead.

rjeczalik committed:
kloud/metadata: workaround cloud-init logging for Red Hat

rjeczalik committed:
kloud/metadata: workaround for group on Suse

rjeczalik committed:
stack/provider: add koding_klient_timeout support

Allows for overwriting default timeout for "Checking remote vm"
phase of building a stack - which is currently 5m.

rjeczalik committed:
kloud/provider: update tests

cihangir commented on go/src/koding/kites/kloud/metadata/metadata.go:
```diff
@@ -1,11 +1,27 @@
+// TODO(rjeczalik): make it configurable during ./configure = add to kites/config/config.json
+var (
+	DefaultEntrypointBaseURL = "https://koding-klient.s3.amazonaws.com/entrypoint"
+	DefaultScreenURL         = "https://koding-dl.s3.amazonaws.com/screen.tar.gz"
```
we could use https://koding.com/d/$1 instead of https://koding-dl.s3.amazonaws.com/$1

rjeczalik replied:
Done.

cihangir commented on go/src/koding/kites/kloud/metadata/metadata.go:
```diff
@@ -1,11 +1,27 @@
+//
+// TODO(rjeczalik): make it configurable during ./configure = add to kites/config/config.json
+var (
+	DefaultEntrypointBaseURL = "https://koding-klient.s3.amazonaws.com/entrypoint"
```
we could utilize koding.com/s3 endpoint  here or we could copy entrypoint to dl bucket and serve from there from now on

rjeczalik replied:
Done, went with /s3.

cihangir commented on go/src/koding/kites/kloud/metadata/provision.sh:
```diff
@@ -5,38 +5,72 @@
+		if [ $${USE_EMBEDDED} -eq 1 ]; then
+			install_screen
+		elif type yum &>/dev/null; then
+			yum install --assumeyes screen
```
could we install a specific and tested version of screen otherwise we might have problems and it would be hard to debug since we wont have any idea on which version is installed. Even on AWS some of the images(same version) has different versions of tools.

rjeczalik replied:
Done.

cihangir commented on go/src/koding/klient/Dockerfile:
```diff
@@ -6,5 +6,5 @@ ENV GOPATH /opt/koding/go
 
 RUN apt-get update && apt-get install -y git build-essential ubuntu-dev-tools dh-make && apt-get clean && \
-    curl -sSL https://storage.googleapis.com/golang/go1.7.4.linux-amd64.tar.gz | tar -C /usr/local/ -zxf - && \
+    curl -sSL https://storage.googleapis.com/golang/go1.8.linux-amd64.tar.gz | tar -C /usr/local/ -zxf - && \
```
could we remove this dockerfile and use koding's base docker image? lets not maintain two images

cc/ @szkl

rjeczalik replied:
FWIW this Dockerfile is going away, once the deb builds are removed.

rjeczalik committed:
provider/marathon: update tests

rjeczalik committed:
provider/softlayer: update tests

rjeczalik committed:
kloud/metadata: workaround missing output

rjeczalik committed:
kloud/metadata: use embedded screen by default for all distros

If terminfo exists on a vm use it instead.

Fixes the following error when starting a session:

  Cannot find terminfo entry for 'xterm-256color'

rjeczalik committed:
stack/provider: remove unused property - koding_use_embedded

rjeczalik committed:
kloud/metadata: reverse proxy s3 files through koding.com

rjeczalik committed:
koding: reformat generated files

rjeczalik committed:
kloud/metadata: do not overwrite cloud-init output

rjeczalik committed:
klient/termina: guess TERM

rjeczalik committed:
provider/azure: update tests

rjeczalik committed:
klient/terminal: add debug logging

rjeczalik committed:
klient/terminal: use xterm-color for ubuntu

rjeczalik commented:
@cihangir All addressed, ptal.

cihangir commented on go/src/koding/klient/terminal/command.go:
```diff
@@ -31,6 +40,22 @@ func init() {
+func guessTerm() string {
+	terms := [][2]string{
+		{"xterm-256color", "/usr/share/terminfo/x/xterm-256color"},
+		{"xterm-256color", "/usr/share/terminfo/78/xterm-256color"},
```
78 is the new answer to life the universe and everything

rjeczalik replied:
😂

rjeczalik replied:
In fact 78 is hex code for `x` - under Linux terminfo files are grouped in directories after the first letter, under macOS - after hex code value of the first letter.
